File: Classes/AudioPlayer.m
//
//  AudioPlayer.m
//  KTF-v1.7
//
//  Created by satya on 3/21/11.
//  Copyright 2011 __MyCompanyName__. All rights reserved.
//

#import "AudioPlayer.h"
#import "KidsTimeFunAppState.h"

static AudioPlayer* audioPlayer = nil;

@implementation AudioPlayer
@synthesize queuePlayer;

+(AudioPlayer*)getInstance
{
	if (audioPlayer == nil) {
		audioPlayer = [[AudioPlayer alloc] init];
	}
	return audioPlayer;
}

-(void)playAudioForHours:(int)hours andMinutes:(int)minutes
{
	NSMutableArray* filesToPlay = [NSMutableArray array];
	NSString* audioFilePath;

	if (minutes == 0)	//If minutes == 0
	{
		if (hours == 12) //If hours == 12
		{
			int randVar = arc4random() % 3;
			switch (randVar) 
			{
				case 0:
					audioFilePath = [[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"12_m"] ofType:@"mp3"];
					[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];
					audioFilePath = [[NSBundle mainBundle] pathForResource:@"oclock_m" ofType:@"mp3"];
					[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];
					break;
				case 1:
					audioFilePath = [[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"12_midnight_m"] ofType:@"mp3"];
					[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];
					break;
				case 2:
				default:
					audioFilePath = [[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"12_noon_m"] ofType:@"mp3"];
					[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];
					break;
			}			
		}
		else //If hours != 12
		{
			audioFilePath = [[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"%d_m",hours] ofType:@"mp3"];
			[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];
            audioFilePath = [[NSBundle mainBundle] pathForResource:@"oclock_m" ofType:@"mp3"];
			[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];
		}
	}
	else if (minutes == 15)
	{
		int randVar = arc4random() % 2;
		switch (randVar) 
		{
			case 0:
				audioFilePath = [[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"%d_m",hours] ofType:@"mp3"];
				[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];
				audioFilePath = [[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"15_m"] ofType:@"mp3"];
				[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];				
				break;
			case 1:
			default:
				audioFilePath = [[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"quarter_past_m"] ofType:@"mp3"];
				[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];								
				audioFilePath = [[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"%d_m",hours] ofType:@"mp3"];
				[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];
				break;
		}
	}
	else if (minutes == 30)
	{
		int randVar = arc4random() % 2;
		switch (randVar) 
		{
			case 0:
				audioFilePath = [[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"%d_m",hours] ofType:@"mp3"];
				[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];
				audioFilePath = [[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"30_m"] ofType:@"mp3"];
				[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];				
				break;
			case 1:
			default:
				audioFilePath = [[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"half_past_m"] ofType:@"mp3"];
				[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];								
				audioFilePath = [[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"%d_m",hours] ofType:@"mp3"];
				[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];
				break;
		}		
	}
	else if (minutes == 45)
	{
		int randVar = arc4random() % 2;
		switch (randVar) 
		{
			case 0:
				audioFilePath = [[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"%d_m",hours] ofType:@"mp3"];
				[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];
				audioFilePath = [[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"45_m"] ofType:@"mp3"];
				[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];				
				break;
			case 1:
			default:
				audioFilePath = [[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"quarter_til_m"] ofType:@"mp3"];
				[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];								
				audioFilePath = [[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"%d_m",hours == 12 ? 1 : hours+1] ofType:@"mp3"];
				[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];
				break;
		}				
	}
	else if (minutes <= 29)
	{
		audioFilePath = [[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"%d_m",minutes] ofType:@"mp3"];
		[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];
		audioFilePath = [[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"minutes_past_m"] ofType:@"mp3"];
		[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];						
		audioFilePath = [[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"%d_m",hours] ofType:@"mp3"];
		[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];
	}
	else 
	{
		int randVar = arc4random() % 2;
		switch (randVar) 
		{
			case 0:
				audioFilePath = [[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"%d_m",minutes] ofType:@"mp3"];
				[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];
				audioFilePath = [[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"minutes_past_m"] ofType:@"mp3"];
				[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];						
				audioFilePath = [[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"%d_m",hours] ofType:@"mp3"];
				[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];
				break;
			case 1:
			default:
				audioFilePath = [[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"%d_m",60-minutes] ofType:@"mp3"];
				[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];
				audioFilePath = [[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"minutes_til_m"] ofType:@"mp3"];
				[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];						
				audioFilePath = [[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"%d_m",hours == 12 ? 1 : hours+1] ofType:@"mp3"];
				[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];
				break;
		}						
	}

	[self playQueuedFiles:filesToPlay];
}

-(void)playAudioTime:(NSString*)time
{
	NSArray* timeComponents = [time componentsSeparatedByString:@":"];
	[self playAudioForHours:[[timeComponents objectAtIndex:0] intValue] andMinutes:[[timeComponents objectAtIndex:1] intValue]];
}

-(void)playQueuedFiles:(NSArray*)fileNames
{
	if (![[KidsTimeFunAppState sharedState] appSoundState]) {
		return;
	}
	
	NSMutableArray* queueFiles = [NSMutableArray array];
	for (NSURL* url in fileNames) {
		AVPlayerItem* playerItem = [AVPlayerItem playerItemWithURL:url];
		[queueFiles addObject:playerItem];
	}
	
	[self.queuePlayer removeAllItems];
	self.queuePlayer = [AVQueuePlayer queuePlayerWithItems:queueFiles];
	[self.queuePlayer play];
}

-(void)audioPlayerDidFinishPlaying:(AVAudioPlayer *)player successfully:(BOOL)flag
{

}

-(void)dealloc
{
	if (audioPlayer) {
		[audioPlayer release];
	}	
	[super dealloc];
}

-(void)playCorrectWrong:(BOOL)correct
{
	int randNum = arc4random() % 9 + 1;
	NSString* soundFile;
	if (correct) 
	{
		soundFile = [NSString stringWithFormat:@"%dcorrect.mp3",randNum];
	}
	else 
	{
		soundFile = [NSString stringWithFormat:@"%dwrong.mp3",randNum];
	}

	NSString* audioFilePath = [[NSBundle mainBundle] pathForResource:soundFile ofType:nil];
	NSMutableArray* filesToPlay = [NSMutableArray array];
	[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];
	[self playQueuedFiles:filesToPlay];
}

-(void)playAudioFile:(NSString*)fileName
{
	NSString* audioFilePath = [[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"%@_m",fileName] ofType:@"mp3"];

	NSMutableArray* filesToPlay = [NSMutableArray array];
	[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];
	[self playQueuedFiles:filesToPlay];	
}

-(void)playAudioFile:(NSString*)fileName withTime:(NSString*)strTime
{
	NSMutableArray* filesToPlay = [NSMutableArray array];
	NSString* audioFilePath = [[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"%@_m",fileName] ofType:@"mp3"];
	[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];
	
	NSArray* timeComponents = [strTime componentsSeparatedByString:@":"];
	
	audioFilePath = [[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"%d_m",[[timeComponents objectAtIndex:0] intValue]] ofType:@"mp3"];
	[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];

	if([[timeComponents objectAtIndex:1] intValue] != 0)
	{
		audioFilePath = [[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"%d_m",[[timeComponents objectAtIndex:1] intValue]] ofType:@"mp3"];
		[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];		
	}
    else
    {
        audioFilePath = [[NSBundle mainBundle] pathForResource:@"oclock_m" ofType:@"mp3"];
        [filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];
    }

	[self playQueuedFiles:filesToPlay];	
}

-(void)playTellTime:(NSString*)fileName playHours:(BOOL)pHrs hours:(NSString*)hrs playMinutes:(BOOL)pMnts minutes:(NSString*)mnts playAnd:(BOOL)sAnd playAgo:(BOOL)sAgo
{
	NSMutableArray* filesToPlay = [NSMutableArray array];
	NSString* audioFilePath = [[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"%@_m",fileName] ofType:@"mp3"];
	[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];
	
	NSArray* tComponens;
	if (pHrs) {
		tComponens = [hrs componentsSeparatedByString:@" "];
		audioFilePath = [[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"%@_m",[tComponens objectAtIndex:0]] ofType:@"mp3"];
		[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];
		audioFilePath = [[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"%@_m",[tComponens objectAtIndex:1]] ofType:@"mp3"];
		[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];		
	}

	if (sAnd) {
		audioFilePath = [[NSBundle mainBundle] pathForResource:@"and_m" ofType:@"mp3"];
		[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];
	}
	
	if (pMnts) {
		tComponens = [mnts componentsSeparatedByString:@" "];
		audioFilePath = [[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"%@_m",[tComponens objectAtIndex:0]] ofType:@"mp3"];
		[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];
		audioFilePath = [[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"%@_m",[tComponens objectAtIndex:1]] ofType:@"mp3"];
		[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];		
	}
	
	if (sAgo) {
		audioFilePath = [[NSBundle mainBundle] pathForResource:@"ago_m" ofType:@"mp3"];
		[filesToPlay addObject:[NSURL fileURLWithPath:audioFilePath]];
	}
	[self playQueuedFiles:filesToPlay];	
}

@end

File: Classes/AudioPlayer.swift
import AVFoundation

class AudioPlayer: NSObject {

    private static let sharedInstance = AudioPlayer()

    class func getInstance() -> AudioPlayer {
        return sharedInstance
    }

    private var queuePlayer: AVQueuePlayer?

    func playAudio(forHours hours: Int, andMinutes minutes: Int) {
        var files: [URL] = []

        func mp3(_ name: String) -> URL? {
            guard let path = Bundle.main.path(forResource: name, ofType: "mp3") else { return nil }
            return URL(fileURLWithPath: path)
        }

        if minutes == 0 {
            if hours == 12 {
                let r = Int.random(in: 0...2)
                switch r {
                case 0:
                    [mp3("12_m"), mp3("oclock_m")].compactMap { $0 }.forEach { files.append($0) }
                case 1:
                    if let u = mp3("12_midnight_m") { files.append(u) }
                default:
                    if let u = mp3("12_noon_m") { files.append(u) }
                }
            } else {
                [mp3("\(hours)_m"), mp3("oclock_m")].compactMap { $0 }.forEach { files.append($0) }
            }
        } else if minutes == 15 {
            if Int.random(in: 0...1) == 0 {
                [mp3("\(hours)_m"), mp3("15_m")].compactMap { $0 }.forEach { files.append($0) }
            } else {
                [mp3("quarter_past_m"), mp3("\(hours)_m")].compactMap { $0 }.forEach { files.append($0) }
            }
        } else if minutes == 30 {
            if Int.random(in: 0...1) == 0 {
                [mp3("\(hours)_m"), mp3("30_m")].compactMap { $0 }.forEach { files.append($0) }
            } else {
                [mp3("half_past_m"), mp3("\(hours)_m")].compactMap { $0 }.forEach { files.append($0) }
            }
        } else if minutes == 45 {
            let nextHour = hours == 12 ? 1 : hours + 1
            if Int.random(in: 0...1) == 0 {
                [mp3("\(hours)_m"), mp3("45_m")].compactMap { $0 }.forEach { files.append($0) }
            } else {
                [mp3("quarter_til_m"), mp3("\(nextHour)_m")].compactMap { $0 }.forEach { files.append($0) }
            }
        } else if minutes <= 29 {
            [mp3("\(minutes)_m"), mp3("minutes_past_m"), mp3("\(hours)_m")].compactMap { $0 }.forEach { files.append($0) }
        } else {
            let nextHour = hours == 12 ? 1 : hours + 1
            if Int.random(in: 0...1) == 0 {
                [mp3("\(minutes)_m"), mp3("minutes_past_m"), mp3("\(hours)_m")].compactMap { $0 }.forEach { files.append($0) }
            } else {
                [mp3("\(60 - minutes)_m"), mp3("minutes_til_m"), mp3("\(nextHour)_m")].compactMap { $0 }.forEach { files.append($0) }
            }
        }

        playQueuedFiles(files)
    }

    func playAudioTime(_ time: String) {
        let parts = time.components(separatedBy: ":")
        guard parts.count == 2,
              let h = Int(parts[0]),
              let m = Int(parts[1]) else { return }
        playAudio(forHours: h, andMinutes: m)
    }

    func playQueuedFiles(_ urls: [URL]) {
        guard KidsTimeFunAppState.sharedState().appSoundState else { return }
        let items = urls.map { AVPlayerItem(url: $0) }
        queuePlayer?.removeAllItems()
        queuePlayer = AVQueuePlayer(items: items)
        queuePlayer?.play()
    }

    func playCorrectWrong(_ correct: Bool) {
        let n = Int.random(in: 1...9)
        let name = correct ? "\(n)correct.mp3" : "\(n)wrong.mp3"
        guard let path = Bundle.main.path(forResource: name, ofType: nil) else { return }
        playQueuedFiles([URL(fileURLWithPath: path)])
    }

    func playAudioFile(_ fileName: String) {
        guard let path = Bundle.main.path(forResource: "\(fileName)_m", ofType: "mp3") else { return }
        playQueuedFiles([URL(fileURLWithPath: path)])
    }

    func playAudioFile(_ fileName: String, withTime time: String) {
        var files: [URL] = []
        func mp3(_ name: String) -> URL? {
            guard let path = Bundle.main.path(forResource: name, ofType: "mp3") else { return nil }
            return URL(fileURLWithPath: path)
        }
        if let u = mp3("\(fileName)_m") { files.append(u) }
        let parts = time.components(separatedBy: ":")
        if parts.count == 2, let h = Int(parts[0]), let m = Int(parts[1]) {
            if let u = mp3("\(h)_m") { files.append(u) }
            if m != 0 {
                if let u = mp3("\(m)_m") { files.append(u) }
            } else {
                if let u = mp3("oclock_m") { files.append(u) }
            }
        }
        playQueuedFiles(files)
    }

    func playTellTime(_ fileName: String, playHours pHrs: Bool, hours hrs: String,
                      playMinutes pMnts: Bool, minutes mnts: String,
                      playAnd sAnd: Bool, playAgo sAgo: Bool) {
        var files: [URL] = []
        func mp3(_ name: String) -> URL? {
            guard let path = Bundle.main.path(forResource: "\(name)_m", ofType: "mp3") else { return nil }
            return URL(fileURLWithPath: path)
        }
        if let u = mp3(fileName) { files.append(u) }
        if pHrs {
            let parts = hrs.components(separatedBy: " ")
            parts.compactMap { mp3($0) }.forEach { files.append($0) }
        }
        if sAnd, let u = Bundle.main.path(forResource: "and_m", ofType: "mp3").map({ URL(fileURLWithPath: $0) }) {
            files.append(u)
        }
        if pMnts {
            let parts = mnts.components(separatedBy: " ")
            parts.compactMap { mp3($0) }.forEach { files.append($0) }
        }
        if sAgo, let u = Bundle.main.path(forResource: "ago_m", ofType: "mp3").map({ URL(fileURLWithPath: $0) }) {
            files.append(u)
        }
        playQueuedFiles(files)
    }
}
